### PR TITLE
Analytics: Track more events

### LINKED ISSUE #855: Matomo track product page activity pt 2
I think https://github.com/iFixit/react-commerce/pull/802 was merged a little prematurely. 

AFAICT, the following aren't registering in Matomo, or at least the way I would expect. Events listed have screenshots of the same event on cominor and how it's recorded there.

<details><summary>Add to cart & Mini cart clicks</summary>

image

Clicking on View cart does register, but not as a mini-cart click. It just looks like a link click?
image

</details>

<details><summary>Add to cart Frequently Bought Together</summary>

image

</details>

<details><summary>Featured Products click</summary>

image

</details>

<details><summary>Newsletter footer submit</summary>

image

</details>

<details><summary>Footer link click (is registering somewhat, but not quite the same as on cominor)</summary>

On cominor, it's getting registered as its own event.
image

On the react pages, it's registering as a sub-event(?)
image

</details>

From what I can tell, these are registering as expected:
- Warranty clicks
- Abandoned cart

Testing details:
<details><summary>Visit 6 on this user profile has the details that were recorded when I performed each of the above actions on the main branch</summary>

https://matomo.ubreakit.com/index.php?date=today&module=Widgetize&action=iframe&visitorId=dd9162939bce75c8&idSite=1&period=day&moduleToWidgetize=Live&actionToWidgetize=getVisitorProfilePopup
image

</details>

cc: @sterlinghirsh @masonmcelvain

### PR DESCRIPTION
**Pull request by masonmcelvain**
This tracks more events in Matomo/GA across Next.js, making it more similar to what we already do for iFixit/ifixit pages.

## CR

Our iFixit/ifixit event tracking implementation: https://github.com/iFixit/ifixit/blob/323d77c499fc79b1088e453da3a17442a3a4369c/carpenter-frontend/Shared/Analytics/CombinedGAMatomo.ts#L29

Search for usages of this function in iFixit/ifixit to find the analogs to the usages here.

## QA

Check for each of the types of events in the parent issue. The wording might not be exactly the same, but make sure we're at least getting the relevant information for each type of event.

You'll need the [google tag manager legacy extension](https://chrome.google.com/webstore/detail/tag-assistant-legacy-by-g/kejbdjndbnbjgmefkgdddjlbokphdefk?hl=en) to QA this.

Click "Record", refresh the page, perform some actions, click "Stop recording", then view the report to see the events.

CC @erinemay

Closes #855

### CODE REVIEW

**Review by sterlinghirsh — DISMISSED** [review]
CR :zap:

**Comment by erinemay** [comment]
Confirmed the following are working as expected now:
- Newsletter submit
- Add to cart
- Mini cart: View cart and Checkout

dev_block 🧷 on a couple of things, and I'll test GA after lunch.

## Matomo
### The frequently bought together cart adds aren't distinguishable from regular adds to cart 
Here's a single add to cart event followed by viewing the mini cart. 
Then I added the two items to my cart via the Frequently Bought Together section.
(For reference, this is [visit 1 on this user's profile](https://matomo.ubreakit.com/index.php?date=today&module=Widgetize&action=iframe&visitorId=999dcc23ebef7802&idSite=1&period=day&moduleToWidgetize=Live&actionToWidgetize=getVisitorProfilePopup). Skip past Jordy's subdomain clicks. We're both working from the office and  I think it's confused that we're both on windows+ff+same ip)
<details><summary>screenshot</summary>

image

</details>

### Featured Product Click isn't distinguishable from other clicks
Here (on the same user profile as above, just a little farther down), I clicked on the Minnow from the Featured Product Section and there isn't any indication that the link was clicked on from that section. I used the back button to get back to the Macbook Retina battery product page after that.
<details><summary>screenshot</summary>

image

</details>

## Other
On this branch preview lots of the footer links are lost. I'm not sure if this was a bug that got fixed and this branch is behind. Could this be from your changes? 
Ref product (that I used for the above testing as well) https://react-commerce-git-matomo-product-page-v2-ifixit.vercel.app/products/macbook-12-retina-early-2015-battery
For example, ⬆️ this link in FF and Chrome on Windows, the footer looks like this:

<details><summary>screenshot</summary>

image

</details>

**Comment by erinemay** [comment]
Also...
### Footer clicks aren't registering as footer clicks
<details><summary>Legacy page</summary>
 
image

</details>

<details><summary>Current branch</summary>

image

</details>

**Comment by masonmcelvain** [comment]
I merged in `main` and made it so "Frequently bought together" add to cart events are distinguishable. I'm heading out for the day so I'll get to the rest of this tonight or tomorrow

File: packages/newsletter-sdk/index.tsx [hunk 1]
```diff
@@ -41,6 +42,10 @@ export function useSubscribeToNewsletter(): [Subscription, SubscribeFn] {
                status: SubscriptionStatus.Subscribed,
                error: undefined,
             }));
+            trackInMatomoAndGA({
+               eventCategory: 'Newsletter',
+               eventAction: `Subscribe Form - Submit`,
```
Comment (masonmcelvain) [inline packages/newsletter-sdk/index.tsx:47]:
```suggestion
               eventAction: 'Subscribe Form - Submit',
```

**Comment by masonmcelvain** [comment]
Thanks for the great QA!

> footer links are lost
> The collection list is gone

I think these were fixed by merging in `main` 🥳 

I made code changes to fix the rest of the problems you observed. un_dev_block ✌🏻

**Comment by erinemay** [comment]
I'm still seeing the missing product list and footer elements on this branch, even with the most recent deploy:
https://react-commerce-apzeaarg2-ifixit.vercel.app/Parts

I'm not seeing it in the prod preview, so I think it's a data issue, but doesn't this use the same data as main? On which this isn't happening. Or does each branch spin up its own strapi instance with its own data?

(Testing the ga/matomo things now. Will add another comment with all that.)

**Comment by masonmcelvain** [comment]
> Or does each branch spin up its own strapi instance with its own data?

Each branch has it's own strapi, which is probably the culprit here. Locally the data looks good to me, which is why I thought it was resolved. You can find the strapi for this branch here https://matomo-product-page-v2.govinor.com/admin

**Review by sterlinghirsh — DISMISSED** [review]
CR :zap:

**Comment by erinemay** [comment]
### Matomo:
All of the following checked ones are registering in Matomo and are clearly identified as that type of event as with our php pages. 
- [x] Newsletter submit
- [x] Mini cart view cart
- [x] Mini cart check out
- [x] Abandoned cart
- [x] Footer links
- [x] Featured Product click
- [ ] Frequently bought together

The frequently bought together cart adds don't say that they are frequently bought together adds. 
image

This is visit 2 of [this user's profile](https://matomo.ubreakit.com/index.php?date=today&module=Widgetize&action=iframe&visitorId=999dcc23ebef7802&idSite=1&period=day&moduleToWidgetize=Live&actionToWidgetize=getVisitorProfilePopup)

Each of these add to carts were done from the frequently bought together section. The first two cart adds in this screenshot were done together. The last one was done individually by deselecting all but that option in the frequently bought together and adding it to cart. 

Product page used: 
https://react-commerce-git-matomo-product-page-v2-ifixit.vercel.app/products/samsung-galaxy-s20-ultra-usa-usb-c-charge-port-genuine

**Comment by erinemay** [comment]
Gonna try GA again after a meeting

**Comment by masonmcelvain** [comment]
The frequently bought together bug should be fixed!

**Review by sterlinghirsh — DISMISSED** [review]
CR :zap:

**Comment by erinemay** [comment]
When I try to run local-dev, I get this error: 
image


And I can view admin at http://localhost:1337/admin, but I get "Refused to connect" when I try http://localhost:3000/Parts

**Comment by erinemay** [comment]
But frequently bought together bug is fixed. So Matomo is good. I just can't get local-dev working to test GA

**Comment by erinemay** [comment]
Actually, I'm getting the same error on main. Gonna try troubleshooting.

**Comment by erinemay** [comment]
@masonmcelvain 
GA - followed setup steps [here](https://github.com/iFixit/react-commerce/pull/594#issue-1330225458) and confirmed the following are registering in google analytics:

I'm a little unsure what's expected here. Do we have a spec sheet on what GA events should look like?
When I did all of the following events, this is what was recorded in the Tag Assistant Legacy extension:
- Newsletter submit
- Mini cart view cart
- Mini cart check out
- Click on a footer link
- Featured Product click
- Frequently bought together add to cart
- Regular add to cart

image

<details><summary>It's also saying there are errors in the tag assistant report</summary>

image

</details>

I would expect more detail than just a recording of links clicked on.

**Comment by masonmcelvain** [comment]
I don't think we have a spec sheet of what the GA events should look like. The best I can suggest is going to [ifixit.com](http://ifixit.com/) and performing the same actions.
The ERR_BLOCKED_BY_CLIENT messages seem like an adblocker or something is preventing those requests from going through, make sure your browser and extensions are all turned off/as loose as possible.
Here is what I saw when I submitted the newsletter form just now

image

**Comment by erinemay** [comment]
Ok, we got it working, but there are duplicate add to cart events.
When I added a single item to cart (the Mako) and then added the three frequently bought together items, these are the add to cart events that got recorded:
image


After examining all of those events, it looks like the frequently purchased together events are duplicated.
<details><summary>We get the expected 3 frequently add</summary>

image

</details>

<details><summary>Then we get the same events recorded as regular add to cart events (with the original single product add at position 2)</summary>

image

</details>

<details><summary>Otherwise, the following are all present:</summary>

Mini cart view cart
image


Mini cart check out
image

Click on a footer link and newsletter
image

Featured Product click
image


</details>

**Comment by masonmcelvain** [comment]
Thanks @erinemay! I've update the code to only fire one event per add to cart action, even if that action included multiple items.

In the screenshot below, the first event was the primary add to cart button, and you can see that only one sku appears in the event.

The second event was the "Frequently Bought Together" add to cart button, and you can see 3 skus in the add to cart event.

image

This is ready to be QA-ed again.

**Review by sterlinghirsh — APPROVED** [review]
CR :zap:

**Comment by erinemay** [comment]
I followed those setup steps and recorded events on the live site and this branch locally performing the following actions:
1. Add to cart
2. Mini cart view cart
3. Mini cart checkout
4. Frequently bought together add to cart
5. Device click
6. Replacement guide click
7. Featured product click
8. Footer click
9. Newsletter subscribe

All of the above actions are getting recorded as events except for the frequently bought together. It's now not being recorded at all when following a regular add to cart. 

This happens when I go to a page like http://localhost:3000/products/macbook-pro-16-2019-battery, click the single (standard) add to cart button followed immediately by clicking the frequently bought together add to cart.

**Comment by erinemay** [comment]
I also noticed that because clicking on a product link from the product page doesn't prompt a full page load, the extension is adding events from a previous page onto that new pageload.

Steps:
1. Go to http://localhost:3000/products/macbook-pro-16-2019-battery
2. Follow recording steps and start recording
3. On http://localhost:3000/products/macbook-pro-16-2019-battery and perform some cart adds
4. Click on a featured product (minnow) at the bottom (notice no full page load)
5. Perform some actions on that new page
6. Stop recording
7. Check the extension's recorded tags and notice that all the events performed on that original page are attributed to the featured page.

View the recorded events and see that all the events done on http://localhost:3000/products/macbook-pro-16-2019-battery are recorded under the minnow pageload.

There is no pageload event for http://localhost:3000/products/macbook-pro-16-2019-battery, only the minnow.
image


## There are two identical tags with identical contents:
image

### Contents of the tags:
<details><summary>Two page views</summary>

image

</details>

<details><summary>Three events</summary>

image

</details>

The page views and events are correct.


The two (potential) problems:
1. The duplicate tags
2. The fact that those tags and all the events under those tags are showing up under that minnow page load

@masonmcelvain, I think the duplication of tags might be a bug.
I'm gonna find @jeffsnyder and see how well this environment translates over to real GA and if the inclusion of everything under the same pageload is relevant.

**Comment by masonmcelvain** [comment]
> except for the frequently bought together. It's now not being recorded at all when following a regular add to cart.

Hm I can't reproduce this behavior (even though I watched you do it!). The following steps produce the correct events for me (with and without the optional step):

- Visit http://localhost:3000/products/iphone-11-replacement-battery
- Start recording
- Click "Add to cart"
- Scroll down and click "Add to cart" in the frequently bought together section
- Optional: Refresh the page
- Stop recording

<details>

image
</details>

**Comment by danielbeardsley** [comment]
> I also noticed that because clicking on a product link from the product page doesn't prompt a full page load, in some cases, the extension is disregarding the original page load and adding all the events done on that page and on the second page under a single page load.

If we haven't already tracked next.js page loads in matomo, we may need something like this:
https://github.com/SocialGouv/matomo-next

Good catch!

**Comment by masonmcelvain** [comment]
@danielbeardsley yea that's a cool library we've talked about using. I think Matomo is pretty squared away in this PR, so I don't really wanna touch it again, but it would be nice to experiment with post launch. The issues @erinemay is seeing are with GA.

**Comment by danielbeardsley** [comment]
> I think Matomo is pretty squared away in this PR

Good deal!

I just glanced at the code and it looks like we're already tracking the router events from next. I just hadn't seen that yet.
https://github.com/iFixit/react-commerce/blob/b25cf4a9541eefda509c1c09def612a9068ec2e2/frontend/components/analytics/Matomo.tsx#L9-L16

**Comment by masonmcelvain** [comment]
> I think the duplication of tags might be a bug.

I searched around online, and it seems like the extension might tell us if we're mistakenly tracking things twice. Here's an example:

<details>



_https://webmasters.stackexchange.com/questions/121076/google-tag-assistant-reports-same-web-property-id-is-tracked-twice_

</details>

I still can't explain why that tag is duplicated. Initially I thought it might be a next/script problem, but https://github.com/vercel/next.js/issues/30984 (and it's [demo](https://external-script-routing.vercel.app/)) seem to indicate that external scripts only get loaded once across page navigations.

**Comment by erinemay** [comment]
>Hm I can't reproduce this behavior (even though I watched you do it!). The following steps produce the correct events for me (with and without the optional step):

There's an error with adding that product to cart. It's another bug. I see the frequently bought together product adds for other products.

**Comment by masonmcelvain** [comment]
Omg so annoying. Glad it works now

**Comment by erinemay** [comment]
dev_block 🧷 @sterlinghirsh says it should show up as separate page loads. The single page load recorded is wrong.

**Comment by danielbeardsley** [comment]
> says it should show up as separate page loads

Yes, but I think we have the code in place to do that, so it makes me suspicious of the tooling used. Maybe the network tab and inspecting `minerva.ifixit.com` requests is our better friend here?

**Comment by masonmcelvain** [comment]
I posted in #data-analytics in slack about this (https://ifixit.slack.com/archives/C01B6G4DV1D/p1666910812808039).

Both @ifixitmaxb and @jeffsnyder aren't entirely sure what the consequences of this will be, but it seems like pageviews will still be in the same session, which is a requirement.

I'm gonna call this QA ✌🏻 and if there are problems in GA we can fix it post launch. All the data will be there at least, it's just a matter of organization at this point.

**Comment by masonmcelvain** [comment]
un_dev_block ✌🏻

**Pull request merged**